File: app-java/src/main/java/com/tritiumstudios/phasmophobiaevidencetool/android/data/viewmodels/MapMenuViewModel.java
package com.tritiumstudios.phasmophobiaevidencetool.android.data.viewmodels;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.TypedArray;

import androidx.annotation.DrawableRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModel;

import com.tritiumstudios.phasmophobiaevidencetool.R;
import com.tritiumstudios.phasmophobiaevidencetool.android.activities.investigation.mapsmenu.data.MapData;
import com.tritiumstudios.phasmophobiaevidencetool.android.activities.investigation.mapsmenu.mapdisplay.data.models.MapModel;

/**
 * MapMenuViewModel class
 *
 * @author TritiumGamingStudios
 */
public class MapMenuViewModel extends ViewModel {

    private Thread imageDisplayThread;

    private MapData[] mapData;
    private int currentMap = 0;

    private MapModel currentMapModel;

    public void init(@NonNull Context context) {

        // SET DATA
        if (!hasMapData()) {
            setMapData(context);
        }

    }

    /**
     * setMapData method
     *
     * @param context
     */
    @SuppressLint("ResourceType")
    public void setMapData(@NonNull Context context) {

        TypedArray typedArray =
                context.getResources().obtainTypedArray(R.array.maps_resources_array);
        MapData[] mapData = new MapData[typedArray.length()];
        for (int i = 0; i < typedArray.length(); i++) {
            mapData[i] = new MapData();

            TypedArray mapTypedArray =
                    context.getResources().obtainTypedArray(typedArray.getResourceId(i, 0));

            //Set map name
            mapData[i].setMapName(mapTypedArray.getString(0));

            //Set map layer names
            TypedArray mapLayerNamesIDs =
                    context.getResources().obtainTypedArray(mapTypedArray.getResourceId(4, 0));
            for (int j = 0; j < mapLayerNamesIDs.length(); j++) {
                mapData[i].addFloorName(mapLayerNamesIDs.getResourceId(j, 0));
            }
            mapLayerNamesIDs.recycle(); //cleanup

            //Set map default layer
            mapData[i].setDefaultFloor(mapTypedArray.getInt(5, 0));

            //Set map thumbnail resource id
            mapData[i].setThumbnail(mapTypedArray.getResourceId(7, 0));

            //Set map layer primary images
            TypedArray mapImages =
                    context.getResources().obtainTypedArray(mapTypedArray.getResourceId(8, 0));
            for (int j = 0; j < mapImages.length(); j++) {
                mapData[i].addFloorLayer(j, mapImages.getResourceId(j, 0));
            }
            mapImages.recycle(); //cleanup

            mapTypedArray.recycle();
        }
        typedArray.recycle();
        setMapData(mapData);
    }

    /**
     * setMapData method
     *
     * @param mapData array
     */
    public void setMapData(MapData[] mapData) {
        this.mapData = mapData;
    }

    /**
     * hasMapData method
     *
     * @return isMapData null
     */
    public boolean hasMapData() {
        return mapData != null;
    }

    /**
     * setMapData method
     *
     * @return mapData array
     */
    public MapData[] getMapData() {
        return mapData;
    }

    @Nullable
    public String[] getMapNames() {
        if(hasMapData()) {
            String[] mapNames = new String[mapData.length];
            for(int i = 0; i < mapNames.length; i++) {
                mapNames[i] = mapData[i].getMapName();
            }
            return mapNames;
        }
        return null;
    }

    public @DrawableRes int[] getMapThumbnails() {
        if(hasMapData()) {
            @DrawableRes int[] mapThumbnails = new int[mapData.length];
            for(int i = 0; i < mapThumbnails.length; i++) {
                mapThumbnails[i] = mapData[i].getThumbnailImage();
            }
            return mapThumbnails;
        }
        return null;
    }

    /**
     * getMapDataLength method
     *
     * @return the size of the mapData array
     */
    public int getMapDataLength() {
        if (!hasMapData()) {
            return 0;
        }
        return mapData.length;
    }

    /**
     * hasCurrentMapData method
     *
     * @return hasMapData
     */
    public boolean hasCurrentMapData() {
        return hasMapData() && currentMap < mapData.length && mapData[currentMap] != null;
    }

    /**
     * setCurrentMapData method
     *
     * @param currentMapPos
     */
    public void setCurrentMapData(int currentMapPos) {
        if (hasMapData() && currentMapPos < mapData.length) {
            this.currentMap = currentMapPos;
        }
    }

    /**
     * getCurrentMapData method
     *
     * @return MapData
     */
    @Nullable
    public MapData getCurrentMapData() {
        if (!hasMapData()) {
            return null;
        }
        return mapData[currentMap];
    }

    /**
     * setImageDisplayThread method
     *
     * @param thread
     */
    public void setImageDisplayThread(Thread thread) {
        imageDisplayThread = thread;
    }

    /**
     * getImageDisplayThread method
     *
     * @return imageDisplayThread
     */
    public Thread getImageDisplayThread() {
        return imageDisplayThread;
    }

    public void setCurrentMapModel(MapModel currentMapModel) {
        this.currentMapModel = currentMapModel;
    }

    public MapModel getCurrentMapModel() {
        return this.currentMapModel;
    }

}

File: app-java/src/main/java/com/TritiumGaming/phasmophobiaevidencepicker/data/viewmodels/MapMenuViewModel.kt
package com.TritiumGaming.phasmophobiaevidencepicker.data.viewmodels

import android.annotation.SuppressLint
import android.content.Context
import android.content.res.Resources
import androidx.annotation.DrawableRes
import androidx.lifecycle.ViewModel
import com.TritiumGaming.phasmophobiaevidencepicker.R
import com.TritiumGaming.phasmophobiaevidencepicker.data.viewmodels.models.maps.map.MapModel
import com.TritiumGaming.phasmophobiaevidencepicker.data.viewmodels.models.maps.mapviewer.MapViewerModel

/**
 * MapMenuViewModel class
 *
 * @author TritiumGamingStudios
 */
class MapMenuViewModel : ViewModel() {

    var imageDisplayThread: Thread? = null

    private var mapsData: MutableList<MapViewerModel> = mutableListOf()

    var currentMapIndex: Int = 0
        set(currentMapPos) {
            if (currentMapPos < mapsData.size) {
                field = currentMapPos
            }
        }
    val currentMapData: MapViewerModel
        get() {
            return mapsData[currentMapIndex]
        }
    fun incrementFloorIndex() {
        var layerIndex: Int = currentMapData.currentFloor
        if (++layerIndex >= currentMapData.floorCount) {
            layerIndex = 0
        }
        currentMapData.currentFloor = layerIndex
    }
    fun decrementFloorIndex() {
        var layerIndex: Int = currentMapData.currentFloor
        if (--layerIndex < 0) {
            layerIndex = currentMapData.floorCount -1
        }
        currentMapData.currentFloor = layerIndex
    }

    var currentMapModel: MapModel? = null

    /*
    val mapNames: MutableList<String>
        get() {
            val mapNames = mutableListOf<String>()
            mapsData.forEachIndexed { index, it ->
                mapNames.add(index, it.mapName)
            }
            return mapNames
        }
    */

    val mapThumbnails: MutableList<Int>
        @DrawableRes get() {
            @DrawableRes val mapThumbnails = mutableListOf<Int>()
            mapsData.forEachIndexed { index, it ->
                mapThumbnails.add(index, it.thumbnailImage)
            }
            return mapThumbnails
        }

    fun init(context: Context) {
        if (mapsData.isEmpty()) buildMapData(context)
    }

    @SuppressLint("ResourceType")
    fun buildMapData(context: Context) {
        val resources: Resources = context.resources
        val allMapsTypedArray =
            resources.obtainTypedArray(R.array.maps_resources_array)

        val tempMapsData: MutableList<MapViewerModel> = mutableListOf()
        for (mapIndex in 0 until allMapsTypedArray.length()) {

            val tempMapData = MapViewerModel()

            val nameKey = 0
            val layerNamesKey = 4
            val defaultLayerKey = 5
            val thumbnailKey = 7
            val layerImagesKey = 8

            val mapTypedArray =
                resources.obtainTypedArray(allMapsTypedArray.getResourceId(mapIndex, 0))

            //Set map name
            tempMapData.mapName = mapTypedArray.getString(nameKey) ?: tempMapData.mapName

            //Set map layer names
            val mapLayerNamesIDs =
                resources.obtainTypedArray(mapTypedArray.getResourceId(layerNamesKey, 0))
            for (j in 0 until mapLayerNamesIDs.length()) {
                tempMapData.floorNames.add(j, mapLayerNamesIDs.getResourceId(j, 0))
            }
            mapLayerNamesIDs.recycle() //cleanup

            //Set map default layer
            tempMapData.defaultFloor = mapTypedArray.getInt(defaultLayerKey, 0)

            //Set map thumbnail resource id
            tempMapData.thumbnailImage = mapTypedArray.getResourceId(thumbnailKey, 0)

            //Set map layer primary images
            val mapImages =
                resources.obtainTypedArray(mapTypedArray.getResourceId(layerImagesKey, 0))
            for (j in 0 until mapImages.length()) {
                tempMapData.addFloorLayer(j, mapImages.getResourceId(j, 0))
            }
            mapImages.recycle() //cleanup
            mapTypedArray.recycle()

            tempMapsData.add(tempMapData) // add to temp maps array
        }
        allMapsTypedArray.recycle()

        this.mapsData = tempMapsData // finaly set maps temp array into maps array
    }

    fun hasCurrentMapData(): Boolean {
        return currentMapIndex < mapsData.size
    }


}
